### PR TITLE
Add support for DIMM temperature sensor

### PR DESCRIPTION
**Pull request by lich426**
I modified the code and did PR again. (#395)

### CODE REVIEW

**Comment by Cubox** [comment]
Hi,

I tested the produced artefacts from the CI build, unfortunately none of my 4 DIMM are showing on this build. They showed in the build of the previous patch (with a ghost 8th stick at 13°C)

**Comment by lich426** [comment]
@Cubox Thank you for your test.
Could you test it again?

**Comment by PhyxionNL** [comment]
The previous version detected two DIMM sensors incorrectly. This one only detects one anymore (which is a bit odd as I have two identical sticks), but it's some progress.

```timeout == MAX_TIMEOUT || (temp & 0x10) > 0 || (temp & 0x08) > 0 || (temp & 0x04) > 0```
Where is this based on? temp for this is "2" for me, maybe ``temp & 0x02 > 0`` needs to be checked as well? 

FWIW:
image

**Comment by xcasxcursex** [comment]
Looks good here 
image

**Comment by lich426** [comment]
@PhyxionNL
I don't know the code that controls the smbus.
The code that controls the smbus refers to the code on the link. (​https://github.com/CalcProgrammer1/OpenRGB)
I think I need a code that excludes fake sensors. but I'm not sure how :(
The datasheets for the sensors are at the link below.

https://www.onsemi.com/pdf/datasheet/cat34ts04-d.pdf
https://www.renesas.com/us/en/document/dst/tse2004gb2b0-datasheet?r=113729
https://www.nxp.com/docs/en/data-sheet/SE97B.pdf

**Comment by Cubox** [comment]
image
Works for me.
Thanks for your work

**Comment by PhyxionNL** [comment]
> 
> 
> @PhyxionNL
> I don't know the code that controls the smbus.
> The code that controls the smbus refers to the code on the link. (​https://github.com/CalcProgrammer1/OpenRGB)
> I think I need a code that excludes fake sensors. but I'm not sure how :(
> The datasheets for the sensors are at the link below.
> 
> https://www.onsemi.com/pdf/datasheet/cat34ts04-d.pdf
> https://www.renesas.com/us/en/document/dst/tse2004gb2b0-datasheet?r=113729
> https://www.nxp.com/docs/en/data-sheet/SE97B.pdf

I checked some of those and I think it should be possible, there should a bit that should be set when the temperature monitor is disabled. There are also temperature resolutions specified depending on two bits.

File: LibreHardwareMonitorLib/Hardware/Memory/IntelDimmSensor.cs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        {
+
+        }
+
+        public override void UpdateSensor()
+        {
+            try
+            {
+                ushort data = GetWord(_address, 0x05);
+                var temp = BitConverter.GetBytes(data);
```
Comment (zeule) [inline LibreHardwareMonitorLib/Hardware/Memory/IntelDimmSensor.cs:26]: Errr.... Why not simply:
```cs
// MSB -> LSB
byte upper = (byte)(data & 0x1F);
byte lower = (byte)(data >> 8);
```
?

File: LibreHardwareMonitorLib/Hardware/Memory/IntelDimmSensor.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            Ring0.WriteSmbus(SMBHSTADD, ((addr & 0x7f) << 1) | SMB_READ);
+            Ring0.WriteSmbus(SMBHSTCMD, command);
+
+            Ring0.WriteSmbus(SMBAUXCTL, Ring0.ReadSmbus(SMBAUXCTL) & (~SMBAUXCTL_CRC));
+
+            Transaction(0x0C);
+
+            Ring0.WriteSmbus(SMBAUXCTL, Ring0.ReadSmbus(SMBAUXCTL) & ~(SMBAUXCTL_CRC | SMBAUXCTL_E32B));
+
+            ushort wordData = (ushort)(Ring0.ReadSmbus(SMBHSTDAT0) + (Ring0.ReadSmbus(SMBHSTDAT1) << 8));
```
Comment (zeule) [inline LibreHardwareMonitorLib/Hardware/Memory/IntelDimmSensor.cs:79]: Redundant variable.

File: LibreHardwareMonitorLib/Hardware/Memory/IntelDimmSensor.cs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        private static int Transaction(byte xact)
+        {
+            int result = CheckPre();
+            if (result < 0)
+                return result;
+
+            Ring0.WriteSmbus(SMBHSTCNT, Ring0.ReadSmbus(SMBHSTCNT) & ~SMBHSTCNT_INTREN);
+            Ring0.WriteSmbus(SMBHSTCNT, xact | SMBHSTCNT_START);
+
+            int status = WaitIntr();
```
Comment (zeule) [inline LibreHardwareMonitorLib/Hardware/Memory/IntelDimmSensor.cs:92]: Redundant variable.

File: LibreHardwareMonitorLib/Hardware/Memory/IntelDimmSensor.cs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            if (timeout > maxCount)
+            {
+                return -1;
+            }
+            return status & (STATUS_ERROR_FLAGS | SMBHSTSTS_INTR);
+        }
+
+        private static int CheckPost(int status)
+        {
+            int result = 0;
```
Comment (zeule) [inline LibreHardwareMonitorLib/Hardware/Memory/IntelDimmSensor.cs:142]: Redundant variable.

File: LibreHardwareMonitorLib/Hardware/Memory/AmdDimmSensor.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        public AmdDimmSensor(string name, int index, Hardware hardware, ISettings settings, byte address) : base(name, index, hardware, settings, address)
+        {
+
+        }
+
+        public override void UpdateSensor()
+        {
+            try
+            {
+                ushort data = GetWord(_address, 0x05);
```
Comment (zeule) [inline LibreHardwareMonitorLib/Hardware/Memory/AmdDimmSensor.cs:24]: What's the point in composing BE word in `GetPort()` if you always switch bytes afterwards?

**Review by zeule — COMMENTED** [review]
Maybe that can help you to simplify the code.

**Comment by PhyxionNL** [comment]
I have not made any other changes, but looking at the documentation the 8th bit in capabilities register says if it's enabled or not. With this change it's no longer detecting fake sensors here, but I'm looking for some feedback on this from users that do have sensors. The temperature resolution would also need to be added as right now it doesn't seem to do that.

**Comment by xcasxcursex** [comment]
> I'm looking for some feedback on this from users that do have sensors

I'm afraid that this most recent change has been the only one so far that didn't work for me. No DIMM temperature sensors appear at all.

**Comment by xcasxcursex** [comment]
That got it working again for me: 
image

**Comment by atiensivu** [comment]
I'm not familiar with downloading from artifacts - I'd like to test this on my PC that has sensors in the DIMMs and provide some feedback. Is there a link to the artifact/pull build?

**Comment by xcasxcursex** [comment]
Click the tickbox next to the commit number, then click Details: 
image

Click Artifacts then grab the approriate build: 
image


But now that this has been merged (yay) you can use any of the CI builds after this commit date, so you can get the latest from the readme.md

**Comment by Snazzie** [comment]
Nice! Works with my corsair vengence pro sticks on gigabyte x570 master .

**Comment by Cubox** [comment]
> But now that this has been merged (yay) you can use any of the CI builds after this commit date, so you can get the latest from the readme.md

It has been merged?

**Comment by tocirahl** [comment]
Is there any news on when this will be merged?

**Comment by PhyxionNL** [comment]
> 
> 
> Is there any news on when this will be merged?

I hope to get around to this eventually, this is a pretty big PR and the code isn't in a great state (some of it has already been commented on) and also doesn't use the the DIMM temperature resolutions (although that can come at a later point).

**Comment by northmatt** [comment]
Will this merge request ever see the light of day? Im hoping that when I upgrade to RaptorLake/Zen4 and Lovelace/RDNA3 that I wont have to merge myself and do bug fixes or be sad that I wouldnt know how to make it read DDR5 data 😢.

**Comment by PhyxionNL** [comment]
> Will this merge request ever see the light of day? Im hoping that when I upgrade to RaptorLake/Zen4 and Lovelace/RDNA3 that I wont have to merge myself and do bug fixes or be sad that I wouldnt know how to make it read DDR5 data 😢.

The above comment still applies. You're free to help in cleaning this up.

**Comment by grzonu** [comment]
bump, any progress here ?

**Comment by PhyxionNL** [comment]
> bump, any progress here ?

You're more than welcome to help with this.

**Comment by PhyxionNL** [comment]
Closing due to stale and conflicts, feel free to open again with the requested changes addressed.

**Comment by Blacktempel** [comment]
Please check #395 for progress state. Would be great to have a couple additional testers providing log files.
Thank you.

**Pull request closed without merging**